src/main/java/java/lang/String.java

```diff
@@ -3884,6 +3884,47 @@ public String toString() {
         return result;
     }
 
+    /**
+     * Helper function for the {@code format} function.
+     * Serialize potential String.format argument to a String.
+     * The returned String is never null.
+     */
+    static String cproverFormatArgument(Object obj)
+    {
+        if(obj == null)
+            return "";
```

Comment:
:question: In the JDK implementation of `String.format`, `null` is not generally formatted as an empty string. For example,
```
    String f = String.format("%s %s", null, "hello");
    System.out.println(f);
```
prints `null hello`. Why does this method return an empty string in this case?

Reply:
https://docs.oracle.com/javase/7/docs/api/java/util/Formatter.html has a "Conversions" section that explains some of the null cases. For `%s` and `%h`, `null` is formatted as `null`, but for `%b`, it is formatted as `false`. It looks like for the character and number cases, null is not an expected input, but with `%c` and null I also got it to print `null`.

Reply:
@antlechner Fixed that to give "null" for null strings

Reply:
I think the special cases will have to be handled in the solver because at this point we don't know which format specifier is used for the corresponding argument.  And for %b %c we will be able to distinguish "null" from valid inputs (which will begin with "000").

Reply:
That makes sense. But I think there is still a problem because sometimes `null` and `"null"` can give different results. For example,
```
String.format("%b", null);
```
is `false` but
```
String.format("%b", "null");
```
is `true`. Is there an easy way to fix this? If not then the limitation should be documented.

Reply:
This could be fixed but I'm afraid the solution I have in mind would make handling the string case a bit more complicated, so the whole would be slower, so I prefer to document the limitation (and I hope giving a String to match a "%b" is rare).

Reply:
@antlechner done

Reply:
I agree it's better for now to document the limitation of an edge case than to slow down the general case.